src/observer.ts

```diff
@@ -0,0 +1,444 @@
+import { Modifiers, StepEvent, MousemoveRecord } from './types';
+import { EventBus, EventHandler, RegisterOptions } from './util/eventbus';
+import { ThrottleManager } from './util/throttler';
+
+type ResetHandler = () => void;
+type EmitHandler = (event: StepEvent, target: HTMLElement | null) => void;
+
+function on(
+  type: string,
+  fn: EventListenerOrEventListenerObject,
+  target: Document | Window,
+  options: AddEventListenerOptions = { capture: true, passive: true },
+): ResetHandler {
+  target.addEventListener(type, fn, options);
+  return () => target.removeEventListener(type, fn, options);
+}
+
+function toModifiers(options: {
+  ctrlKey: boolean;
+  shiftKey: boolean;
+  altKey: boolean;
+  metaKey: boolean;
+}): Modifiers {
+  return {
+    ctrl: options.ctrlKey || undefined,
+    alt: options.altKey || undefined,
+    shift: options.shiftKey || undefined,
+    meta: options.metaKey || undefined,
+  };
+}
+
+export type ObserverOptions = {
+  doc: Document;
+  win: Window;
+  onEmit: EmitHandler;
+  onMouseDown?: EventHandler;
+  onMouseUp?: EventHandler;
+  onClick?: EventHandler;
+  onMouseMove?: EventHandler;
+  onScroll?: EventHandler;
+  onKeyDown?: EventHandler;
+  onKeyPress?: EventHandler;
+  onKeyUp?: EventHandler;
+  onTextInput?: EventHandler;
+  onBlur?: EventHandler;
+};
+
+export const EventSymbol = {
+  mouseDown: Symbol('mousedown'),
+  mouseUp: Symbol('mouseup'),
+  click: Symbol('click'),
+  mouseMove: Symbol('mousemove'),
+  scroll: Symbol('scroll'),
+  keyDown: Symbol('keydown'),
+  keyPress: Symbol('keypress'),
+  keyUp: Symbol('keyup'),
+  textInput: Symbol('textinput'),
+  blur: Symbol('blur'),
+};
+
+export class EventObserver {
+  private win: Window;
+  private doc: Document;
+  private onEmit: (
+    event: StepEvent,
+    target: HTMLElement | null,
+    fromThrottler?: boolean,
+  ) => void;
+  private handlers: ResetHandler[] = [];
+
+  private active = false;
+  private started = false;
+
+  private throttleMananger = new ThrottleManager();
+
+  private eventBus = new EventBus();
+
+  constructor(options: ObserverOptions) {
+    const {
+      win,
+      doc,
+      onEmit,
+      onMouseDown,
+      onMouseUp,
+      onClick,
+      onMouseMove,
+      onScroll,
+      onKeyDown,
+      onKeyPress,
+      onKeyUp,
+      onTextInput,
+      onBlur,
+    } = options;
+    this.win = win;
+    this.doc = doc;
+    this.onEmit = (event, target, fromThrottler = false) => {
+      !fromThrottler && this.throttleMananger.invokeAll();
```

Comment:
`!fromThrottler` means flush all the events in a throttler?

Reply:
Yes, if a event which not come from throttler is going to emit, we need to emit all the events throtting to keep the correct  order of the event chain.